README.md
# Featured Users

Allows the administrator to make users featured. All it does is give the user a custom meta field called `jsfeatured_user`. Now available are a shortcode, widget, filters, and theme functions.

## Description

While working on a custom WordPress theme we needed the ability to feature users and authors. So we created this ‘Featured users’ plugin which allows the administrator to easily star (feature) users. Then using a custom query in the theme, shortcode or widget; featured users can be displayed by only retrieving users with the custom user meta `jsfeatured_user`. The plugin does not make any changes to your current theme it only does the following:

* Adds a featured column in the Users panel of the WordPress Admin.
* When a user is featured the plugin adds or updates the custom field ‘jsfeatured_user’ and sets the value to yes.
* Provides custom widget that echoes out shortcode.
* Provides custom shortcode that echoes ul list of featured users.
* Includes theme functions and filters for developers.

Feel free to use and include in your WordPress Installs, please think of us if you need a custom theme or plugin developed! Use documented code examples on the settings page to modify the expected output.

## Installation

1. Visit 'Plugins > Add New'
1. Search for 'Featured Users'
1. Click on 'Install Now'
1. Activate the Featured Users plugin.
1. Go to 'Settings > Featured Users' and modify.

## How to

After activating and after updating the plugin's settings go to wp-admin > Users > All Users and feature a few users. To use the 'Featured Users' plugin edit a page and add the 'Featured Users' shortcode or update a sidebar and add the 'Featured Users' widget.

## Shortcode Paramaters

[rd-featured-users role="administrator,author" avatar="yes" max="100"]

* role = A comma-separated list of roles of the users you want to include in the query. Note the settings page can override this. The default is all roles again unless the settings page specifies which to allow.
* avatar = Options are yes or no, specify whether to show or not show the users' avatar image. The default is no.
* max = The maximum number of users to display.

## Filters

WordPress filters allow a 'developer' to modify aspects of the 'Featured Users' plugin without editing the core plugin. If you are not familiar with filters [please read this](https://codex.wordpress.org/Plugin_API) before continuing.

1. 'featured-users-args', [https://codex.wordpress.org/Class_Reference/WP_User_Query](https://codex.wordpress.org/Class_Reference/WP_User_Query)
1. 'featured-users-css', url to css file
1. 'featured-users-JS', url to js file
1. 'featured-user-shortcode-row', user row in shortcode and widget
1. 'featured-user-shortcode-return', shortcode and widget content

## Frequently Asked Questions

Let us know what questions you have!

## Support

For custom WordPress plugin and theme development requests email us at [info@reactivedevelopment.net](mailto:info@reactivedevelopment.net) or go to [https://www.reactivedevelopment.net/](https://www.reactivedevelopment.net/). If you have questions or requests for this plugin go to [http://wordpress.org/support/plugin/featured-users-wordpress-plugin](http://wordpress.org/support/plugin/featured-users-wordpress-plugin) or for quick and paid support go to [https://www.reactivedevelopment.net/contact/project-mind/?plugin=featured-users](https://www.reactivedevelopment.net/contact/project-mind/?plugin=featured-users) to message us.

changelog.txt
= Changelog archive =

= 2.1 =

Release Date: 2019/01/14

* Updated settings page content, see inc/settings.php. Updated readmes as well.

= 2.0 =

Release Date: 2018/12/31

* Finished 2.0 development, testing and documentation.

= 1.6 =

Release Date: 2018/12/17

* Added new functions rd_is_user_featured() and rd_featured_users()

= 1.5 =

Release Date: 2018/12/17

* Re-developed by, Jeremy Selph http://www.reactivedevelopment.net/							

= 1.0 =

* Released to WordPress.

= 0.1 =

* initial plugin development by, Jeremy Selph http://www.reactivedevelopment.net/

inc/widget.php
<?php /**
 *
 * @package Featured Users
 * @subpackage widget.php
 * @author Jeremy Selph <jselph@reactivedevelopment.net>
 * @link http://www.reactivedevelopment.net/
 * @license GPLv3
 * @version 0.1

    Change Log

     * 0.1 	initial plugin development by, Jeremy Selph http://www.reactivedevelopment.net/ 		    2018/12/17

    Plugin Class

     * @package rd_featured_Users_widget
     * @author Jeremy Selph <jselph@reactivedevelopment.net>
     * @version 0.1
     * @access public
    */
	class rd_featured_Users_widget extends WP_Widget {

        /**
         * @package __construct
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function __construct(){            
            parent::__construct(

                'rd_featured_Users_widget', 
                'Featured users',  
                array(

                    'classname' => 'rd_featured_Users_widget',
                    'description' => 'This widget displays all of your featured users in a one-column list.'

                )

            );
        }
        
        /**
         * @package widget
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function widget( $args, $instance ){

            if ( !isset( $args['widget_id'] ) ) { $args['widget_id'] = $this->id; }    
            $title = ( ! empty( $instance['title'] ) ) ? $instance['title'] : __( 'Featured Users' );
            $title = apply_filters( 'widget_title', $title, $instance, $this->id_base );

            /**
             * get options from widget
             * @author Jeremy Selph <jselph@reactivedevelopment.net>
             * @since 0.1
            */
            $roles = ( ! empty( $instance['roles'] ) ) ? $instance['roles'] : '';
            $avatar = ( ! empty( $instance['avatar'] ) ) ? $instance['avatar'] : '';
            $max = ( ! empty( $instance['max'] ) ) ? $instance['max'] : '';
    
            /**
             * echo out widget code
             * @author Jeremy Selph <jselph@reactivedevelopment.net>
             * @since 0.1
            */
            echo $args['before_widget'];
            if ( $title ){ echo $args['before_title'] . $title . $args['after_title']; }
            echo do_shortcode( '[rd-featured-users role="' . $roles . '" avatar="' . $avatar . '" max="' . $max . '"]' );
            echo $args['after_widget'];

        }
        
        /**
         * @package form
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function form( $instance ){
            
            $title = ! empty( $instance['title'] ) ? $instance['title'] : ''; ?>
            <p class="feat_users_p">
                <label class="feat_users_title" for="<?php echo $this->get_field_id( 'title' ); ?>"><?php _e( 'Widget title' ); ?>:</label>
                <input class="feat_users_title" type="text" id="<?php echo $this->get_field_id( 'title' ); ?>" name="<?php echo $this->get_field_name( 'title' ); ?>" value="<?php echo esc_attr( $title ); ?>" />
            </p>

    <?php   $featuredUsers = rdFeaturedUsers::getInstance();
            $roles = ! empty( $instance['roles'] ) ? explode( ',', $instance['roles'] ) : array(); ?>
            <p class="feat_users_p">
                <label class="feat_users_roles" for="<?php echo $this->get_field_id( 'roles' ); ?>"><?php _e( 'Roles of users to display' ); ?>:</label>
                <select class="feat_users_roles" multiple="multiple" name="<?php echo $this->get_field_id( 'roles' ); ?>[]" id="<?php echo $this->get_field_id( 'roles' ); ?>">
        <?php   $editable_roles = array_reverse( get_editable_roles() );
                foreach ( $editable_roles as $role => $details ){
                    
                    $key = esc_attr( $role );
                    if( $featuredUsers->is_role_allowed( $key ) ){
                    
                        $name = translate_user_role( $details['name'] );
                        $checked = ( in_array( $key, $roles ) ) ? ' selected="selected"' : '';
                        echo '<option value="' . $key . '"' . $checked . '>' . $name . '</option>';
                    
                    }

                } ?>
                </select>
            </p>
            
    <?php   $avatar = ! empty( $instance['avatar'] ) ? $instance['avatar'] : ''; ?>
            <p class="feat_users_p">
                <label class="feat_users_avatar" for="<?php echo $this->get_field_id( 'avatar' ); ?>"><?php _e( 'Include avatar images' ); ?>:</label>
                <select class="feat_users_avatar" id="<?php echo $this->get_field_id( 'avatar' ); ?>" name="<?php echo $this->get_field_name( 'avatar' ); ?>">
                    <option<?php if( $avatar == 'No' ){ ?> selected="selected"<?php } ?>>No</option>
                    <option<?php if( $avatar == 'Yes' ){ ?> selected="selected"<?php } ?>>Yes</option>
                </select>
            </p>
    
    <?php   $max = ! empty( $instance['max'] ) ? $instance['max'] : ''; ?>
            <p class="feat_users_p">
                <label class="feat_users_max" for="<?php echo $this->get_field_id( 'max' ); ?>"><?php _e( 'Max users to query' ); ?>:</label>
                <input class="feat_users_max" type="text" id="<?php echo $this->get_field_id( 'max' ); ?>" name="<?php echo $this->get_field_name( 'max' ); ?>" value="<?php echo esc_attr( $max ); ?>" />
            </p>
            <p class="feat_users_p"><?php _e( 'This widget displays all of your featured users in a one-column list' ) ?>.</p>
    
    <?php }
        
        /**
         * @package update
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function update( $new_instance, $old_instance ){

            $instance = $old_instance;
            $new_roles = array();
            $instance[ 'title' ] = strip_tags( $new_instance[ 'title' ] );
            $instance[ 'avatar' ] = strip_tags( $new_instance[ 'avatar' ] );
            $instance[ 'max' ] = intval( $new_instance[ 'max' ] );
            
            $roles = $_REQUEST[ $this->get_field_id( 'roles' ) ];
            if( $roles && count( $roles ) > 0 ){
                foreach( $roles as $r ){ 
                    if ( !in_array( $r, $new_roles ) ){ $new_roles[] = sanitize_text_field( $r ); }
                }
            } $instance[ 'roles' ] = implode( ',', $new_roles );
            
            return $instance;

        }        

    } 
    
    /**
     * @package featuredUsersWidget
     * @author Jeremy Selph <jselph@reactivedevelopment.net>
     * @version 0.1
     * @access public
    */
	class featuredUsersWidget extends rdFeaturedUsers {
        
        /**
         * @package __construct
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function __construct(){

            parent::__construct( true );
            
			/**
             * Actions/Filters
             * @author Jeremy Selph <jselph@reactivedevelopment.net>
             * @since 0.1
            */
            add_action( 'widgets_init', array( $this, 'shortcode_content' ) );
            add_action( 'admin_head', array( $this, 'add_widget_css' ) );

        }
        
        /**
         * @package shortcode_content
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function shortcode_content( $atts ){
            
            register_widget( 'rd_featured_Users_widget' );

        }

	    /**
		 * @package add_widget_css
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 0.1
	     * @access public
	    */
	    public function add_widget_css(){

            global $pagenow;
            if ( $pagenow == 'widgets.php' ){
                
              ?><style type="text/css">
                    
                    p.feat_users_p{ display: block; clear: both; margin: 1em 0 1em 0; }
                    input.feat_users_title, select.feat_users_roles, select.feat_users_avatar, input.feat_users_max{ 
                        float: right; width: 50%; }
                    select.feat_users_roles, select.feat_users_avatar{ width: 50%; }
                    label.feat_users_title, label.feat_users_roles, label.feat_users_avatar, label.feat_users_max{ float left; 
                        /* display: block; */ clear: both; display: inline-block; padding: 0.5em 0 0 0; }

                    @media only screen and (max-width: 700px) {

                        label.feat_users_title, label.feat_users_roles, label.feat_users_avatar, label.feat_users_max,
                        input.feat_users_title, select.feat_users_roles, select.feat_users_avatar, input.feat_users_max{
                            float: left; width: 100%; }
                        input.feat_users_title, select.feat_users_roles, select.feat_users_avatar, input.feat_users_max{
                            margin: 0 0 5px 0; }

                    }

                </style><?php

            }
        
        }

    } $featuredUsersWidget = new featuredUsersWidget();

inc/shortcode.php
<?php /**
 *
 * @package Featured Users
 * @subpackage shortcode.php
 * @author Jeremy Selph <jselph@reactivedevelopment.net>
 * @link http://www.reactivedevelopment.net/
 * @license GPLv3
 * @version 0.1

    Change Log

     * 0.1 	initial plugin development by, Jeremy Selph http://www.reactivedevelopment.net/ 		    2018/12/17

    Filters to hook into

	 * 01. 	FILTER : featured-user-shortcode-row [user row in shortcode and widget]
	 * 02. 	FILTER : featured-user-shortcode-return [shortcode and widget content]

    Plugin Class

     * @package featuredUsersShortcode
     * @author Jeremy Selph <jselph@reactivedevelopment.net>
     * @version 0.1
     * @access public
    */
	class featuredUsersShortcode extends rdFeaturedUsers {

        /**
         * @package __construct
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function __construct(){

            parent::__construct( true );
            
			/**
             * Actions/Filters
             * @author Jeremy Selph <jselph@reactivedevelopment.net>
             * @since 0.1
            */
            add_shortcode( 'rd-featured-users', array( $this, 'shortcode_content' ) );

        }
        
        /**
         * @package shortcode_content
         * @author Jeremy Selph <jselph@reactivedevelopment.net>
         * @version 0.1
         * @access public
        */
        public function shortcode_content( $atts ){
            
            $return = '';
            $atts = shortcode_atts(

                array( 'role' => '', 'avatar' => false, 'max' => 1000 ), 
                $atts, 
                'rd-featured-users'

            );
            
            /**
             * Query featured users
             * @author Jeremy Selph <jselph@reactivedevelopment.net>
             * @since 0.1
            */
            $user_query = $this->return_featured_users( $atts[ 'role' ], $atts[ 'avatar' ], $atts[ 'max' ] );
            $users = $user_query->results;
            if ( $users && count( $users ) > 0 ){
                
                $return .= '<ul>';
                foreach( $users as $u ){                    
                    $return .= '<li>';
                        $return .= '<a href="' . get_author_posts_url( $u->ID, get_the_author_meta('user_nicename') ) . '">';
                            
                            if ( strtolower( $atts[ 'avatar' ] ) == 'yes' ){ $return .= get_avatar( $u->ID. 16 ) . ' '; }
                            $return .= apply_filters(

                                'featured-user-shortcode-row',
                                get_the_author_meta( 'display_name', $u->ID ) . ' (' . count_user_posts( $u->ID ) . ')',
                                $u->ID
                                
                            );
                        
                        $return .= '</a>';
                    $return .= '</li>';
                } $return .= '</ul>';
            
            } return apply_filters( 'featured-user-shortcode-return', $return, $users );

        }

    } $featuredUsersShortcode = new featuredUsersShortcode();

js_featured_users.php
<?php /**
 *
 * Plugin Name:   	Featured Users
 * Plugin URI:    	http://www.reactivedevelopment.net/featured-users
 * Description:   	Adds the ability to set a user's custom meta field called "jsfeatured_user".
 * Version:       	2.1
 * Author:        	Jeremy Selph @ Reactive Development LLC
 * Author URI:    	http://www.reactivedevelopment.net/
 * License:       	GPL v3
 * License URI:   	http://www.gnu.org/licenses/gpl-3.0.en.html
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.

	Credits

	 * Plugin Development by Jeremy Selph @ Reactive Development LLC
	 * Star image is from: http://www.iconfinder.com/icondetails/9662/24/bookmark_star_icon and 
	 * http://www.iconfinder.com/icondetails/9604/24/star_icon
 
 	Change Log
	
	 * 2.1 	Updated settings page content, see inc/settings.php. Updated readmes as well. 				2019/01/14
	 * 2.0	Finished 2.0 development, testing and documentation. 										2018/12/31
	 * 1.6 	Added new functions rd_is_user_featured() and rd_featured_users() 							2018/12/17
	 * 1.5 	Re-developed by, Jeremy Selph http://www.reactivedevelopment.net/							2018/12/17
	 * 1.0 	Released to WordPress.
	 * 0.1 	initial plugin development by, Jeremy Selph http://www.reactivedevelopment.net/

	Activation Instructions

	 * 01. 	Visit 'Plugins > Add New'
	 * 02. 	Search for 'Featured Users'
	 * 03.	Click on 'Install Now'
	 * 04. 	Activate the Featured Users plugin.
	 * 05. 	Go to 'Settings > Featured Users' and modify.

    Shortcodes

	 * 01. 	[rd-featured-users role="{comma separated roles of users to include}" avatar="yes|no" max="int"]

    Theme functions

	 * 01. 	rd_is_user_featured( (int)user_id )
	 * 02. 	rd_featured_users( (int)max_users )

    Filters to hook into

	 * 01. 	FILTER : featured-users-args [https://codex.wordpress.org/Class_Reference/WP_User_Query]
	 * 02.	FILTER : featured-users-css [url to css file]
	 * 03.	FILTER : featured-users-JS [url to js file]
	 * 04. 	FILTER : [inc/shortcode.php] featured-user-shortcode-row [user row in shortcode and widget]
	 * 05. 	FILTER : [inc/shortcode.php] featured-user-shortcode-return [shortcode and widget content]

    Plugin Class

     * @package rdFeaturedUsers
     * @author Jeremy Selph <jselph@reactivedevelopment.net>
     * @version 1.5
     * @access public
    */
	class rdFeaturedUsers {

		/**
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5 
    	 * @access public
	    */
		public $_option_name = 'rd-featured-users';

		/**
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5 
    	 * @access public
	    */
		public $_meta_name = 'jsIs_user_featured';

		/**
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5 
    	 * @access public
	    */
		public $_meta_value = 'yes';

		/**
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5 
    	 * @access protected
	    */
		protected $_settings = false;

		/**
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5 
    	 * @access public
	    */
		public $_plugin_url = false;

		/**
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5 
    	 * @access private
	    */
		private $_plugin_dir = false;

		/**
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5 
    	 * @access protected
	    */
		protected $_plugin_base = false;

		/**
		 * Static property to hold class instance
		 * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
    	 * @var rdFeaturedUsers
    	 * @static
	    */
		static $_instance = false;

		/**
		 * @package __construct
    	 * @author Jeremy Selph <jselph@reactivedevelopment.net>
    	 * @version 1.5
    	 * @access public
	    */
		public function __construct( $child = false ){

			$this->_plugin_url = plugin_dir_url( __FILE__ );
			$this->_plugin_dir = plugin_dir_path( __FILE__ );
			$this->_plugin_base = plugin_basename( __FILE__ );
				
			/**
			 * Populate settings
			 * @author Jeremy Selph <jselph@reactivedevelopment.net>
			 * @since 1.5
			*/
			$this->_settings = get_option( $this->_option_name );

			/**
			 * only do the following when the parent class is loaded
			 * @author Jeremy Selph <jselph@reactivedevelopment.net>
			 * @since 1.5
			*/
			if ( !$child ){

				/**
				 * Actions/Filters
				 * @author Jeremy Selph <jselph@reactivedevelopment.net>
				 * @since 1.5
				*/
				add_action( 'show_user_profile', array( $this, 'add_checkBox_to_profile' ), 10, 1 );
				add_action( 'edit_user_profile', array( $this, 'add_checkBox_to_profile' ), 10, 1 );
				add_action( 'personal_options_update', array( $this, 'profile_update_featured_meta' ), 10, 1 );
				add_action( 'edit_user_profile_update', array( $this, 'profile_update_featured_meta' ), 10, 1 );
				add_filter( 'manage_users_columns', array( $this, 'add_featured_column' ), 10, 1 );
				add_filter( 'manage_users_custom_column', array( $this, 'featured_column_content' ), 10, 3 );
				add_action( 'admin_enqueue_scripts', array( $this, 'add_featured_column_css_js' ), 10, 1 );
				add_action( 'wp_ajax_featured_users', array( $this, 'ajax_save_featured_user' ) );
				add_filter( 'plugin_row_meta', array( $this, 'support_link' ), 10, 2 );
				
				/**
				 * Addons
				 * @author Jeremy Selph <jselph@reactivedevelopment.net>
				 * @since 1.5
				*/
				require_once( $this->_plugin_dir . 'inc/shortcode.php' );
				require_once( $this->_plugin_dir . 'inc/widget.php' );

				/**
				 * Addons for admin area
				 * @author Jeremy Selph <jselph@reactivedevelopment.net>
				 * @since 1.5
				*/
				if ( is_admin() ){ require_once( $this->_plugin_dir . 'inc/settings.php' ); }

			}

		}

		/**
		 * @package support_link
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function support_link( $plugin_meta, $plugin_file ){
			if ( $this->_plugin_base === $plugin_file ) {

				$plugin_meta[] = sprintf(
					'&#36; <a href="%s" target="_blank">%s</a>',
					'https://www.reactivedevelopment.net/contact/project-mind/?plugin=featured-users',
					esc_html__( 'Paid support' )
				);

				$plugin_meta[] = sprintf(
					'&hearts; <a href="%s" target="_blank">%s</a>',
					'https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=RESFMU9LDAEDQ&source=url',
					esc_html__( 'Donate' )
				);

			} return $plugin_meta;
		}

	    /**
		 * @package is_user_featured
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function is_user_featured( $user = false ){
			if ( $user ){

				$user_id = ( is_object( $user ) && isset( $user->ID ) ) ? intval( $user->ID ) : intval( $user );
				if ( $user_id > 0 ){ return get_user_meta( $user_id, $this->_meta_name, true ); }

			} return false;
		}

	    /**
		 * @package return_featured_users
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function return_featured_users( $roles = false, $avatar = 'no', $max = 1000 ){
			
			$allowed_roles = false;
			$roles = ( $roles && !is_array( $roles ) && strpos( $roles, ',' ) ) ? explode( ',', $roles ) : array( $roles );
			$user_args = array( 
					
				'meta_key' => $this->_meta_name,
				'meta_value' => $this->_meta_value,
				'number' => $max
				
			);
			
			/**
			 * confirm the roles given are allowed by our settings
			 * @author Jeremy Selph <jselph@reactivedevelopment.net>
			 * @version 1.5
			*/
			if ( $roles && count( $roles ) > 0 ){

				$allowed_roles = array();
				foreach( $roles as $r ){
					if ( $this->is_role_allowed( $r ) ){ $allowed_roles[] = $r; }
				}
				
			}

			/**
			 * if no roles were given and we have roles in our settings lets use them
			 * @author Jeremy Selph <jselph@reactivedevelopment.net>
			 * @version 1.5
			*/
			$allowed_roles = ( !$allowed_roles ) ? $this->allowed_roles() : $allowed_roles;
			if ( count( $allowed_roles ) > 0 ){ $user_args[ 'role__in' ] = $allowed_roles; }

			/**
			 * query the db for featured users
			 * @author Jeremy Selph <jselph@reactivedevelopment.net>
			 * @version 1.5
			*/
			return new WP_User_Query( apply_filters( 'featured-users-args', $user_args ) );

		}

	    /**
		 * @package allowed_roles
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access private
	    */
	    private function allowed_roles(){
			if ( isset( $this->_settings[ 'feat_role' ] ) && !empty( $this->_settings[ 'feat_role' ] ) ){

				return $this->_settings[ 'feat_role' ];

			} return array();
		}

	    /**
		 * @package is_role_allowed
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function is_role_allowed( $roles = array() ){
			
			$roles = ( !is_array( $roles ) ) ? array( $roles ) : $roles;
			$allowed_roles = $this->allowed_roles();
			$role_allowed = ( count( $allowed_roles ) > 0 ) ? false : true;			
			if ( !$role_allowed && count( $roles ) > 0 ){
				foreach( $roles as $r ){
					if ( in_array( $r, $allowed_roles ) ){
						
						$role_allowed = true;
					
					}
				}
			} return $role_allowed;

		}

	    /**
		 * @package save_featured_users
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access private
	    */
	    private function save_featured_users( $users = false ){
			
			$return = $user_ids = array();
			$user_ids = ( !empty( $users ) && !is_array( $users ) ) ? array( $users ) : $users;
			if ( $user_ids && count( $user_ids ) > 0 ){
				
				/**
				 * Loop through each user and update their user meta
				 * @author Jeremy Selph <jselph@reactivedevelopment.net>
				 * @since 1.5
				*/
				foreach( $user_ids as $u ){

					$id = intval( $u );
					$user_meta = $role_allowed = false;
					if ( $id > 0 ){
						
						$user_meta = get_userdata( $id );
						$role_allowed = $this->is_role_allowed( $user_meta->roles );
						if ( $role_allowed ){
						
							update_user_meta( $id, $this->_meta_name, $this->_meta_value );
							$return[] = $id;

						}
					
					}

				} return $return;
				

			} return false;

		}

	    /**
		 * @package delete_featured_user
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access private
	    */
	    private function delete_featured_user( $user = false ){
			
			$user_id = ( is_object( $user ) && isset( $user->ID ) ) ? intval( $user->ID ) : intval( $user );
			if ( $user_id > 0 ){
				
				delete_user_meta( $user_id, $this->_meta_name );
				return true;
			
			} return false;

		}

	    /**
		 * @package ajax_save_featured_user
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function ajax_save_featured_user(){
			if ( wp_verify_nonce( $_REQUEST[ 'nounce' ], 'featured-users' ) &&
				 current_user_can( 'edit_user' ) ){

				$user_id = intval( $_REQUEST[ 'user_id' ] );
				$featured = ( !empty( $_REQUEST[ 'featured' ] ) && $_REQUEST[ 'featured' ] == 'yes' ) ? true : false;
				if ( $user_id > 0 ){
					
					if( $featured ){ $this->save_featured_users( array( $user_id ) ); echo 'added'; }
					else { $this->delete_featured_user( $user_id ); echo 'deleted'; }
				
				}

			} exit;
		}

	    /**
		 * @package profile_update_featured_meta
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function profile_update_featured_meta( $user_id ){
				
			if ( !current_user_can( 'edit_user', $user_id ) ){ return false; }			
			if ( isset( $_REQUEST[ '_' . $this->_meta_name ] ) && 
				 $_REQUEST[ '_' . $this->_meta_name ] == $this->_meta_value ){ $this->save_featured_users( $user_id ); } 
			else { $this->delete_featured_user( $user_id ); }
			
		}

	    /**
		 * @package add_checkBox_to_profile
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function add_checkBox_to_profile( $user ){
			
			$isFeatured	= $this->is_user_featured( $user );
			if ( $this->is_role_allowed( $user->roles ) ){ ?>
			<h3><?php _e( 'Feature User' ); ?></h3>
			<table class="form-table">
		
				<tr>
					<td>
						<input id="_<?php echo $this->_meta_name; ?>" name="_<?php echo $this->_meta_name; 
							?>" type="checkbox" value="<?php echo $this->_meta_value; ?>"<?php if ( $isFeatured ){ ?> checked="checked"<?php } ?>>
						<label for="_<?php echo $this->_meta_name; ?>"><?php _e( 'Check if Featured' ); 
							?> (<span class="description"><?php _e( 'Checked' ); ?> == <?php _e( 'This user is featured' ); 
								?>.</span>)</label>
					</td>
				</tr>
		
			</table><br /><?php
			}
			
		}

	    /**
		 * @package add_featured_column
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function add_featured_column( $columns ){			
			
			$columns[ 'featured_user' ] = __( 'Featured' );
			return $columns;
			
		}

	    /**
		 * @package featured_column_content
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function featured_column_content( $empty, $column_name, $user_id ){
			if( $column_name == 'featured_user' ){
				
				$user_meta = get_userdata( $user_id );
				$role_allowed = $this->is_role_allowed( $user_meta->roles );
				if ( $role_allowed ){
						
					$class =  ( $this->is_user_featured( $user_id ) ) ? ' selected' : '';
					return  '<a id="userFeatured_'.$user_id.'" href="#" class="featuredStar'.$class.'"></a>';

				}
					
			}
		}

	    /**
		 * @package add_featured_column_css_js
	     * @author Jeremy Selph <jselph@reactivedevelopment.net>
		 * @version 1.5
	     * @access public
	    */
	    public function add_featured_column_css_js( $hook ){
		 	if ( $hook == 'users.php' ){

				/**
                 * add admin styles
                 * @author Jeremy Selph <jselph@reactivedevelopment.net>
                 * @version 1.5
                */
                wp_enqueue_style( 
					
					'featured-users-css',
					apply_filters( 'featured-users-css', plugin_dir_url( __FILE__ ) . 'assets/featured_users.css' ),
					array(),
					'0.1'
				
				);

				/**
                 * add admin scripts
                 * @author Jeremy Selph <jselph@reactivedevelopment.net>
                 * @version 1.5
                */
                wp_enqueue_script( 
					
					'featured-users-js', 
					apply_filters( 'featured-users-js', plugin_dir_url( __FILE__ ) . 'assets/featured_users.js' ),
					array( 'jquery' ),
					'0.1', 
					true
				
				);

                /**
                 * add ajax params
                 * @author Jeremy Selph <jselph@reactivedevelopment.net>
                 * @version 1.5
                */				
				wp_localize_script(
					
					'featured-users-js',
					'featured_users',
					array(

						'ajax_url' => admin_url( 'admin-ajax.php' ),
                        'ajaxNounce' => wp_create_nonce( 'featured-users' )

					)

				);				

		 	}
		}

		/**
		 * @package get_class_instance
    	 * @author Jeremy Selph <jselph@reactivedevelopment.net>
    	 * @version 1.5
    	 * @access public
    	 * @return rdFeaturedUsers
	    */
		public static function getInstance(){

			if ( !self::$_instance ){ self::$_instance = new self; } return self::$_instance;

		}

	} $featuredUsers = rdFeaturedUsers::getInstance();

	/**
	 * @package rd_is_user_featured
	 * @author Jeremy Selph <jselph@reactivedevelopment.net>
	 * @version 1.6
	 * @access public
	*/
	function rd_is_user_featured( $user_id = 0 ){

		global $featuredUsers; 
		return $featuredUsers->is_user_featured( $user_id );

	}

	/**
	 * @package rd_featured_users
	 * @author Jeremy Selph <jselph@reactivedevelopment.net>
	 * @version 1.6
	 * @access public
	*/
	function rd_featured_users( $max = 1000 ){
		
		global $featuredUsers; 
		return $featuredUsers->return_featured_users( $max );
	
	}